### ISSUE TITLE
Decimals shift on different bridge sides

### ISSUE THREAD

**Issue by akolotov** [poanetwork/poa-bridge-contracts#243]
As part of https://github.com/poanetwork/tokenbridge/pull/160 it was done a first attempt to introduce the decimals shift for the `erc20-to-native` mode.

It was discussed there that the feature should be implemented on the contract side as the oracle is not intended to perform any modification of the relayed messages.

Therefore it is suggested to introduce a new configuration parameter `FOREIGN_TO_HOME_DECIMAL_SHIFT` which will be used by the bridge contracts to mint/unlock proper amount of tokens. The value of the parameter is the number of decimals required to adjust the amount of tokens.

E.g. `FOREIGN_TO_HOME_DECIMAL_SHIFT=6` means that if a user sends tokens with the value `10^12` on the foreign side, the amount to be passed to mint function on the home side will be `10^18`. And vice versa, if the value of the asset sent to the bridge on the home side is `10^18`, the amount of the token to be unlocked is `10^12`.

Default value for the parameter is 0.

The parameter `FOREIGN_TO_HOME_DECIMAL_SHIFT` will be initialized together with other bridge parameters by `initialize()` method.

It is used by methods `onExecuteMessage` and `onExecuteAffirmation`. Let's consider the logic of the value adjustment for `erc20-to-native` mode:

```solidity
function onExecuteAffirmation(address _recipient, uint256 _value, bytes32 txHash) internal returns(bool) {
    setTotalExecutedPerDay(getCurrentDay(), totalExecutedPerDay(getCurrentDay()).add(_value));
    IBlockReward blockReward = blockRewardContract();
    require(blockReward != address(0));
    uint256 valueToMint = _value * (10 ** decimalShift());

    address feeManager = feeManagerContract();
    if (feeManager != address(0)) {
        uint256 fee = calculateFee(valueToMint, false, feeManager, FOREIGN_FEE);
        distributeFeeFromAffirmation(fee, feeManager, txHash);
        valueToMint = valueToMint.sub(fee);
    }
    blockReward.addExtraReceiver(valueToMint, _recipient);
    return true;
}
```

```solidity
function onExecuteMessage(address _recipient, uint256 _amount, bytes32 _txHash) internal returns(bool) {
    setTotalExecutedPerDay(getCurrentDay(), totalExecutedPerDay(getCurrentDay()).add(_amount));
    uint256 amount = _amount // (10 ** decimalShift());
    return erc20token().transfer(_recipient, amount);
}
```

Since this functionality could be required for other bridge modes it is required to consider this feature interaction with the Fee Manager of any type.

**Comment by branciard** [comment]
Thank you @akolotov for the specification of this feature! 
I [start](https://github.com/iExecBlockchainComputing/poa-bridge-contracts/commit/92ee9a42c9555c82ac3198b5cc724291d21d3626) implementation for the erc_to_native mode. I will continue for the 2 others mode : erc_to_erc and native_to_erc.  

For Fee Manager you mentioned, from my understanding of the [code](https://github.com/poanetwork/poa-bridge-contracts/blob/master/contracts/upgradeable_contracts/BaseFeeManager.sol#L15), currently, it only supports 18 decimals usage, for foreign fees or home fees? Do I treat also the decimals support for fees contracts in this issue ?

**Comment by akolotov** [comment]
@branciard thanks for looking this after!

The code you are referring is to calculate the percentage of the fee. It is not good place to deal with support of different decimals on different bridge sides.

The main difficulty with the supporting decimals shift for fees calculation is that fees distributed differently depending on the bridge mode and fee manager configuration.

For example, the `erc-to-native` mode assumes that the Fee Manager is enabled only on the Home side (which has low gas price) as so it serves transfers in both direction:
- for the direction `Foreign-to-Home`: the value is passed as is by the bridge oracles, then the amount of fees calculated and the value to mint new coins is got as reduction the initial value by the fees amount. The fees are distributed among the validators in coins.
- for the direction `Home-to-Foreign`: the fees amount is calculated and affects the value to be signed by the validators, as soon as enough signatures collected the fee distributed among the validators.

Having this, it seems that nothing is needed to be changed (besides changes described [above](https://github.com/poanetwork/poa-bridge-contracts/issues/243#issue-472480926)) to handle interaction with the Fee Manager:
- `onExecuteAffirmation` will shift decimals as so the fees will be calculated in respect to value in the correct coins amount.
- `onExecuteMessage` does not deal with fees at all - fees will be taken in coins on the Home side, the value will be reduced by amount of fees and then will be adjusted by `onExecuteMessage`.

Similar reasoning could be applied to other Fee Managers.

**Comment by branciard** [comment]
Thank you @akolotov for explanation on Fee Managers.
I submit a proposal PR for this decimal shift feature using contracts instead of shifting in bridges oracles. I remain available on this topic if necessary.

**Comment by akolotov** [comment]
@branciard, the issue is being closed since it was addressed in the release [`3.0.0`](https://github.com/poanetwork/poa-bridge-contracts/releases/tag/3.0.0)
